### ISSUE TITLE
Flow volume generated by a gravity system increases after pumping

### ISSUE THREAD

**Issue by fenclmar** [scheidan/SPG#1]
Gravity systems and pumps were defined as in the example file (line 1 - 130). Then I tried to generate flow from the first gravity system and pumped it. The flow volume after pumping increased for about 7 % compared to flow volume form the system. Where did the extra volume come in?

sim.dur <- 10
temp.res <- 1  
flow.S0 <- G1.S1(sim.dur, temp.res) 
flow.S1 <- P2(flow.S0)
sum(flow.S0[ ,1])
sum(flow.S1[ ,1])

**Comment by scheidan** [comment]
This should not happen... I'd guess that the error is in the internal function `disp.trans`, because the flow distance after the pump has a huge influence on the error:
```R
library(SPG)

set.seed(123)
diurnal.variation.Q.S <- c(4, 8, 15, 20, 12, 10, 11, 7, 6, 10, 8, 5)

G1.S1 <- def.grav(population=40000,
                  frac.consumers=0.005,
                  pulse.mass=5,
                  diurnal.S.WD=diurnal.variation.Q.S,
                  diurnal.S.WE=diurnal.variation.Q.S,
                  Q.c.day=160,
                  diurnal.Q=diurnal.variation.Q.S,
                  Q.base=0,
                  flow.distance=3000)

P2 <- def.pump(V.max=200,
               V.min=1,
               pump.rate=100,
               distance=3000,
               v.flow=1,
               Disp=0.16)

## tests

sim.dur <- 10
temp.res <- 1

flow.S0 <- G1.S1(sim.dur, temp.res)
flow.S1 <- P2(flow.S0)
flow.S2 <- SPG:::disp.trans(flow.S1, distance = 1000, v.flow = 1, 
                            Disp = 0.16)

V0 <- sum(flow.S0[ ,1])
V1 <- sum(flow.S1[ ,1])
V2 <- sum(flow.S2[ ,1])

## relative errors
(V1-V0)/V0*100
(V2-V1)/V1*100                          # error depends strongly in "distance"
```

**Comment by fenclmar** [comment]
Even when distance was zero (to my understanding no diffusion should then occur) the error was significant. The error decreased with increased pump rate and changed with temporal resolution. There might be some problem with starting/stopping the pumping as the pump sump is not filled/depleted continuously.

**Comment by fenclmar** [comment]
I have fixed a small bug in the pump.trans function which was not considering last record of inflow, however, this cannot explain this issue with increased flow after pumping.

**Comment by fenclmar** [comment]
Yes, there is clearly some bug in the disp.trans function

``` R
# define flow
Q <- rep(10, 100) + rnorm(100,0,1)
Q <- data.frame('Q' = Q, 'S' = Q * .01)
attr(Q, "temp.res.sim") <- 60

# plot flow
ramp <- colorRamp(c("red", "white"))
coli <- rgb( ramp(seq(0, 1, length = 11)), max = 255)
plot(Q[ ,1], type='l', ylim = c(0, 21), lwd = 4)
vol_error <- rep(NA, 11) # volume error
for(i in 0:10){
    len <- 2^i
    Qout <- disp.trans(Q, len)
    lines(Qout[ ,1], col = coli[i], lwd = 2) # plot flow
    vol_error[i] <- (sum(Qout[,1]) - sum(Q[,1]))/sum(Q[,1]) # relative error
}
```

Unfortunately, I do not understand the method with FF transform so I do not have clue how to fix it.

**Comment by scheidan** [comment]
Thanks for the example. I'll try to have a look next week

**Comment by scheidan** [comment]
Should be fixed by 301d33b.

```R
## define flow
flow <- c(rep(10, 5), rep(0, 50))  # we must add zeros, the water needs time to travel through the pipe
flow <- as.matrix(data.frame('Q' = flow, 'S' = flow * .01))
attr(flow, "temp.res.sim") <- 60

## test different length
par(mfrow=c(2,1))
plot(flow[ ,1], type='l', ylim = c(0, 21), lwd = 4)
vol_error <- rep(NA, 10) # volume error
for(i in 1:10){
    len <- 2^(i-1)
    flowout <- disp.trans(flow, len)
    lines(flowout[ ,1], col = gray(0.3), lwd = 2) # plot flow
    vol_error[i] <- log(sum(flowout[,1]) / sum(flow[,1]))
    
}

## plot
plot(2^(0:9), vol_error, type="h")
abline(h=0, col=2)

```
